raft/raft.go

```diff
@@ -160,6 +160,10 @@ type Config struct {
 	// overflowing that sending buffer. TODO (xiangli): feedback to application to
 	// limit the proposal rate?
 	MaxInflightMsgs int
+	// MaxUncommittedEntries limits the max number of uncommitted entries that
+	// may be appended to a leader's log. Once this limit is exceeded, proposals
+	// will begin to return ErrProposalDropped errors. Note: 0 for no limit.
+	MaxUncommittedEntries int
```

Comment:
Basing this off the aggregate entry size is tricky. It means we'll need to keep track of the uncommitted entry size as entries are proposed and committed and probably also compute this on leadership changes, which would effectively undo https://github.com/etcd-io/etcd/pull/9073. 

Also, for what it's worth, other flow control mechanisms like `MaxInflightMsgs` operate off message count instead of message size, so there is precendence for this approach.

Reply:
Well, if the size of the log were bounded, #9073 wouldn't have been necessary (or at least not as critical). Also, an approximation here is reasonable: only track the uncommitted log size within a single term; assume that the uncommitted size is zero on becoming leader.

I don't feel too strongly about this given the precedent of MaxInflightMsgs, but if I could go back in time I think I'd change that one to count bytes too.

Reply:
I made the change, but I don't know how I feel about it. I think it's actually harder to pick a reasonable limit based on size than it is is based on entry count. We want the limit to catch runaway proposals without ever affecting normal operation. What size do we set this to to ensure that? A multiple of the max proposal size? An arbitrary large value like 128MB? If we do that then the log can still grow to millions of small entries. A smaller value is no better because then we start dropping large proposals pretty quickly. 

Maybe the argument is that a million small entries is fine as long as they remain under a certain byte size.

Reply:
Yes, my argument is that a million small entries are fine. 

You haven't pushed your branch with this change.

Reply:
Done. PTAL.